---
title: "Andersen Lab Strains"
author: "Dan Cook"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: united
    toc: yes
  pdf_document: 
    toc: yes
---

```{r, echo = F, warning=FALSE, message=FALSE}
library(readr)
library(tidyr)
library(dplyr)
library(knitr)
library(ggplot2)
library(scales)
library(gdata)
library(waffle)
library(DT)
library(ggmap)
library(pryr)
library(lubridate)


pub_theme <- ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(size=14, face="bold", color="black"),
                 axis.text.y = ggplot2::element_text(size=14, face="bold", color="black"),
                 axis.title.x = ggplot2::element_text(size=14, face="bold", color="black", vjust=-.3),
                 axis.title.y = ggplot2::element_text(size=14, face="bold", color="black", vjust=2),
                 strip.text.x = ggplot2::element_text(size=16, face="bold", color="black"),
                 strip.text.y = ggplot2::element_text(size=16, face="bold", color="black"),
                 axis.ticks= element_line( color = "black", size = 0.25), 
                 legend.position="none",
                 plot.margin = unit(c(1.0,0.5,0.5,1.0),"cm"),
                 plot.title = ggplot2::element_text(size=24, face="bold", vjust = 1),
                 legend.position="none",
                 panel.background = ggplot2::element_rect(color = "black",size=0.75),
                 strip.background = ggplot2::element_rect(color = "black", size = 0.75)) 

load_eav <- function(file) {
  readr::read_tsv(file, c("ENTITY", "SUB", "VALUE", "DATE")) %>%
    eav_attr("ATTR") %>%
    dplyr::select(ENTITY, ATTR, SUB, VALUE, DATE)
}

eav_attr <- function(.data, ..., spread = FALSE) {
  columns <- pryr::named_dots(...)
  for(col in columns) {
  .data <- .data %>% tidyr::extract_("SUB", into = c(col), regex = paste0(col, "=","([^;]+)"), remove = FALSE, convert = TRUE)
  }
  if (spread == TRUE) {
    .data <- .data %>% dplyr::select(-DATE, -SUB) %>%
    tidyr::spread(ATTR, VALUE)
  }
  .data
}

seqs <- readr::read_tsv("raw/seq.raw.tsv", na = c("","NA","#N/A"))

na_to_0 <- function(x) {
  ifelse(is.na(x), 0, x)
}

summarize_column <- function(colnames) {
       seqs %>%
       dplyr::filter(processed == TRUE) %>%
       dplyr::group_by_(.dots = colnames, "use_in_analysis") %>%
       dplyr::summarize(sum = n()) %>% 
       dplyr::ungroup() %>%
       tidyr::spread(use_in_analysis, sum) %>%
       dplyr::rename("YES" = `TRUE`, "NO" = `FALSE`) %>%
       dplyr::mutate_each(funs(na_to_0), everything()) %>%
       dplyr::mutate(use_rate = YES / (NO + YES))
}

run_count <- summarize_column("run")
lib_count <- summarize_column(c("run","lib"))
strain_count <- summarize_column(c("run","lib","strain","isotype"))

total_strains <- (strain_count %>%
                 dplyr::filter(use_rate > 0) %>%
                 dplyr::distinct(strain) %>%
                 dplyr::ungroup() %>%
                 dplyr::summarize(n = n()))$n

total_isotypes <- (strain_count %>%
                 dplyr::filter(use_rate > 0) %>%
                 dplyr::distinct(isotype) %>%
                 dplyr::ungroup() %>%
                 dplyr::summarize(n = n()))$n

total_sequences <- (dplyr::filter(seqs, use_in_analysis == TRUE) %>%
                   dplyr::summarize(n_reads = comma(sum(as.numeric(reads), na.rm = TRUE))))$n_reads

total_fastq_pairs <- nrow(dplyr::filter(seqs, use_in_analysis == TRUE))

total_file_size <- dplyr::filter(seqs, use_in_analysis == TRUE) %>%
                    dplyr::mutate(filesize = fq1_size + fq2_size) %>% 
                    dplyr::summarize(humanReadable(sum(as.numeric(filesize))))
        

libraries <- dplyr::group_by(seqs, lib) %>% dplyr::summarize(sum = n())

```

#### Statistics

* __Total Strains__: `r total_strains`
* __Total Isotypes__: `r total_isotypes`
* __Total Sequences__: `r total_sequences`
* __Total Fastq Pairs__: `r total_fastq_pairs`
* __Total File Size__: `r total_file_size`

### Strain Locations

```{r, echo = F, fig.width = 9.5, fig.height = 5, warning=FALSE, message=FALSE}
library(leaflet)

width <- 50
height <- 50

icons <- iconList(
red = makeIcon(
  iconUrl = "red.png",
  iconWidth = width, iconHeight = height,
  popupAnchorX = 0.00001, popupAnchorY = -height+13,
  iconAnchorX = width/2, iconAnchorY = height), 
grey = makeIcon(
  iconUrl = "grey.png",
  iconWidth  = width, iconHeight = height, 
  popupAnchorX = 0.00001, popupAnchorY = -height+13,
  iconAnchorX = width/2, iconAnchorY  = height
))


strain_info <- readr::read_tsv("processed/strain_isotype.tsv") %>%
               dplyr::filter(!is.na(latitude)) %>%
               dplyr::left_join(seqs) %>%
               dplyr::mutate(type = ifelse(!is.na(isotype), "red","grey")) 


attach(strain_info)
leaflet(data = strain_info) %>% 
  addTiles() %>%
  addMarkers(~longitude, ~latitude, popup = paste0(strain, " - ", country), 
             icon = ~icons[type] )


```

<img src="red.png" height="20x" width="20x" /> - Sequenced
<img src="grey.png" height="20x" width="20x" /> - In collection (not sequenced)


### Depth of Coverage

```{r, fig.width = 24, fig.height = 10, echo = F}

doc_global <- load_eav("raw/eav.doc.tsv") %>%
dplyr::filter(ATTR == "depth_of_coverage") %>%
eav_attr("chrom") %>%
dplyr::filter(ATTR == "depth_of_coverage") %>%
dplyr::filter(chrom == "genome")

ggplot(doc_global) +
  geom_bar(aes(x = reorder(ENTITY, VALUE), y = VALUE), stat="identity") +
  geom_hline(aes(yintercept = 30), color = "red") +
  theme_bw() +
  scale_y_continuous(expand = c(0,0), breaks = seq(0, 600, 100), limits = c(0,600)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(x = "Isotype", y = "Depth of Coverage") 

doc_display <- dplyr::select(doc_global, isotype = ENTITY, depth_of_coverage = VALUE) 

write_tsv(doc_display, path = "processed/depth_of_coverage.tsv")

```

<font style="color:red">---</font> depth of coverage = 30

`r datatable(doc_display)`

### Runs

`r kable(run_count)`

```{r, echo = F}

runs <- table(c(dplyr::filter(seqs, use_in_analysis == TRUE) %>% dplyr::select(run)))

colors=c("#882E72", "#B178A6", "#D6C1DE", "#1965B0", "#5289C7", "#7BAFDE", "#4EB265", "#90C987", "#CAE0AB", "#F7EE55", "#F6C141", "#F1932D", "#E8601C", "#DC050C")

waffle(runs, rows=30, size=0.5, 
       colors=rev(colors))
```
n = `r sum(runs)`


### Libraries

`r kable(lib_count)`

```{r generate isotype file, echo = F, warning = F, message = F}

# Output clean seqs file
analysis_seqs <- dplyr::filter(seqs, use_in_analysis == T) %>%
                 dplyr::arrange(run, lib, isotype, strain)

readr::write_tsv(analysis_seqs, path = "processed/seqs.tsv")

isotypes <- seqs %>% 
  dplyr::select(isotype) %>%
  dplyr::filter(!is.na(isotype)) %>%
  dplyr::distinct() %>%
  dplyr::arrange(isotype) %>%
  dplyr::mutate(isotype_count=row_number()) %>%
  dplyr::select(isotype_count, isotype) 

strain_isotype <- seqs %>% 
  dplyr::filter(use_in_analysis == T) %>%
  dplyr::select(isotype, strain, previous_names) %>%
  dplyr::filter(!is.na(isotype)) %>%
  dplyr::distinct() %>%
  dplyr::mutate(sequenced = T)

strain_isolation_information <- readr::read_tsv("https://docs.google.com/spreadsheets/d/1V6YHzblaDph01sFDI8YK_fP0H7sVebHQTXypGdiQIjI/pub?output=tsv")

strain_isotype <- readr::read_tsv("processed/additional_isotypes.tsv") %>%
  dplyr::bind_rows(strain_isotype) %>%
  dplyr::group_by(strain) %>%
  dplyr::mutate(previous_names = ifelse(!is.na(previous_names) & previous_names == strain, NA, previous_names)) %>%
  dplyr::mutate(sequenced = ifelse(is.na(sequenced), F, sequenced)) %>%
  dplyr::left_join(strain_isolation_information) %>%
  dplyr::select(strain, isotype, previous_names, warning_msg, sequenced, everything(),  -location)  %>%
  dplyr::distinct()

strain_isotype$strain_count <- group_indices(strain_isotype) 


readr::write_tsv(isotypes, "processed/isotypes.tsv",  na = "NA")
readr::write_tsv(strain_isotype, "processed/strain_isotype.tsv",  na = "NA")

# Generate Sample File
sample_file <- dplyr::filter(seqs, use_in_analysis == T) %>%                  dplyr::select(fq1, fq2, readgroup, isotype) %>%                   tidyr::separate(readgroup, into = c("RUN", "LB", "STRAIN"), sep ="-",  extra = "drop", remove = F) %>%
  dplyr::mutate(PL = "") %>%
  dplyr::select(FQ1 = fq1, FQ2 = fq2, ID = readgroup, LB, SM = isotype, PL) %>%
  dplyr::arrange(SM)

readr::write_tsv(sample_file, "processed/sample_file.tsv")

# Fastq stats
fastq <- load_eav("raw/fastq.tsv.gz") %>%
         dplyr::select(-SUB, -DATE) %>%
         tidyr::spread(ATTR, VALUE, convert = TRUE) %>%
         dplyr::select(fastq = ENTITY, A_count, T_count, C_count, G_count, everything()) %>%
         #dplyr::mutate_each(funs(type.convert(., as.is=T)), everything()) %>%
         tidyr::separate(fastq, sep = "-", into = c("run", "lib", "strain"), extra = "drop") %>%
         dplyr::group_by(run, lib, strain) %>%
         dplyr::mutate(GC_count_mean = mean(GC_count)) %>%
         dplyr::left_join(analysis_seqs, .)

readr::write_tsv(fastq, path = "processed/fastq_info.tsv")


readr::write_tsv(fastq, "processed/fastq.tsv")

telseq <- readr::read_tsv("raw/telseq.txt") %>%
         dplyr::mutate(LENGTH_ESTIMATE = (as.double(LENGTH_ESTIMATE) / (332720800/46000)) * (5808700/12000)) %>%
         dplyr::filter(TEL15 > 0) %>%
         dplyr::select(-Library, -Duplicates) %>%
         tidyr::separate(ReadGroup, sep = "-", into = c("run", "lib", "strain"), extra = "drop") %>%
         dplyr::group_by(isotype) %>%
         dplyr::mutate(telomere_length = weighted.mean(LENGTH_ESTIMATE, Total))  %>%
         dplyr::select(isotype, run, lib, strain, Total, Mapped, telomere_length, everything(), -Sample)

readr::write_tsv(telseq, "processed/telseq.tsv")



```

### Concordance

Calculate Concordance among strains to determine isotypes.

```{r, echo = F}

df <- readr::read_tsv("raw/concordance_strain.tsv") %>%
      dplyr::mutate(isotype = (concordance > 0.9993))

strain_list <- sort(unique(c(df$sample_i, df$sample_j)))

iso_set <- dplyr::filter(df, isotype == T, sample_i != "LSJ1", sample_j != "LSJ1")

single_strains <- strain_list[!(strain_list %in% unique(c(iso_set$sample_i, iso_set$sample_j)))]
single_strains <- as.data.frame(list(sample_i = single_strains, sample_j = single_strains, discordance = 0, number_of_sites = 100, concordance = 1))

iso_set <- dplyr::bind_rows(iso_set, single_strains) %>%
           dplyr::arrange(sample_i, sample_j)

stack_list <- function(x) {
  if (is.null(names(x)) == T) {
    names(x) <- as.character(1:length(x))
  }
  stack(x)
}

iso_groups <- stack_list(unique(lapply(strain_list, function(x) {
  grouped_strains <- dplyr::filter(iso_set, (sample_i == x | sample_j == x)) %>%
  dplyr::select(sample_i, sample_j)
  sort(unique(unlist(grouped_strains)))
}))) %>%
  dplyr::rename(iso_group = ind, strain = values)  
 
```

`r datatable(iso_groups)`

```{r, echo = F, fig.width = 12, fig.height = 6}


ggplot(df) +
  geom_histogram(aes(x=concordance, fill = isotype), binwidth = 0.0007) +
  scale_fill_manual(values = c("#808080", "#0080FF")) +
  labs(x = "Concordance", y = "Number of Comparisons") +
  geom_vline(aes(xintercept = 0.997), color = "red") +
  scale_y_continuous(expand = c(0,0), limits = c(0,250)) +
  theme(axis.title = ggplot2::element_text(size=14, face="bold", color="black", vjust=5)) +
  pub_theme

```
